styx-common/src/main/java/com/spotify/styx/util/ShardedCounter.java

```diff
@@ -296,35 +258,18 @@ public long getCounter(String counterId) {
    * <p>Due to Datastore limitations (modify max 25 entity groups per transaction),
    * deletion of shards is done in batches of 25 shards.
    *
-   * <p>Behaviour is non-determined if other instances try to access the same counter in the meantime.
+   * <p>Behaviour is best-effort and non-determined if other instances try to access the same counter in the meantime.
    * Best results are achieved if all usages of the given resource - which the counter is associated with -
    * are removed before calling deleteCounter. This is so, in order to avoid a scenario where one instance
    * is trying to delete all shards, while another is creating/updating shards in between the
    * multiple transactions made by this method.
    */
-  public void deleteCounter(Datastore datastore, String counterId) {
-    QueryResults<Entity> results = datastore.run(EntityQuery.newEntityQueryBuilder()
-                                                     .setKind(KIND_COUNTER_SHARD)
-                                                     .setFilter(PropertyFilter
-                                                                    .eq(PROPERTY_COUNTER_ID,
-                                                                        counterId))
-                                                     .build());
-    while (results.hasNext()) {
-      // remove max 25 entities per transaction
-      datastore.runInTransaction(transaction -> {
-        IntStream.range(0, 25).forEach(i -> {
-          if (results.hasNext()) {
-            transaction.delete(results.next().getKey());
-          }
-        });
-        return null;
-      });
+  public void deleteCounter(Storage storage, String counterId) {
+    storage.deleteShardsForCounter(counterId);
+    try {
+      storage.deleteLimitForCounter(counterId);
```

Comment:
Curious why the deletions are differently handled, one has a catch and the other doesn't?

Reply:
Hm, yeah, it does look strange a bit. I think my reasoning for it was that if we fail when deleting some shards we should short-circuit, forward any exception to the caller and keep the limit if it exists, so that it still applies on the remaining shards. And if deleting the shards is successful then try to delete the limit as well.

However, reading the comment now, we do explicitly say that the whole operation is best effort, so maybe we should catch all exceptions after all. What do you think?

Reply:
Discrepancy also in terms of storeWithRetries, and in terms of where the code is located (deleteShardsForCounter is implemented in DatastoreStorage, yet deleteLimitForCounter forwards to an implementation in DatastoreStorageTransaction).

Reply:
Yes, the reason for that is that deleting of a limit can be done in a single transaction, whereas deleting the shards usually takes more than one given the constraint that Datastore transactions have of modifying max 25 entries at once.

Additionally, at the time being, there's no clear contract of what belongs where when it comes to operations against datastore and the two locations of implementations that we have which you mentioned (DatastoreStorage and DatastoreStorageTransaction)